
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <title>Examples &#8212; VariantWorks SDK 0.1.0 documentation</title>
    <link rel="stylesheet" href="_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="API" href="apidocs/modules.html" />
    <link rel="prev" title="Core Features" href="features.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="apidocs/modules.html" title="API"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="features.html" title="Core Features"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">VariantWorks SDK 0.1.0 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="examples">
<h1>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h1>
<p>VariantWorks can be used to setup training and inference pipelines using the available encoders, data loaders
and I/O libraries.</p>
<div class="section" id="snp-zygosity-predictor">
<h2>SNP Zygosity Predictor<a class="headerlink" href="#snp-zygosity-predictor" title="Permalink to this headline">¶</a></h2>
<p>This example showcases how to develop a SNP zygosity predictor from variant SNP candidates. This example is
also available as a sample.</p>
<p>For more details on training and inference using NeMo (including features such as multi-GPU training, half-precision
trainig, model parallel training, etc.), please refer to the <a class="reference external" href="https://nvidia.github.io/NeMo/tutorials/examples.html">Nemo documentation</a>.</p>
<div class="section" id="training">
<h3>Training<a class="headerlink" href="#training" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import nemo and variantworks modules</span>
<span class="kn">import</span> <span class="nn">nemo</span>
<span class="kn">from</span> <span class="nn">variantworks.dataloader</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">variantworks.io.vcfio</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">variantworks.networks</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">variantworks.sample_encoders</span> <span class="kn">import</span> <span class="o">*</span>

<span class="c1"># Create neural factory</span>
<span class="n">nf</span> <span class="o">=</span> <span class="n">nemo</span><span class="o">.</span><span class="n">core</span><span class="o">.</span><span class="n">NeuralModuleFactory</span><span class="p">(</span>
    <span class="n">placement</span><span class="o">=</span><span class="n">nemo</span><span class="o">.</span><span class="n">core</span><span class="o">.</span><span class="n">neural_factory</span><span class="o">.</span><span class="n">DeviceType</span><span class="o">.</span><span class="n">GPU</span><span class="p">,</span> <span class="n">checkpoint_dir</span><span class="o">=</span><span class="n">tempdir</span><span class="p">)</span>

<span class="c1"># Create pileup encoder by selecting layers to encode. More encoding layers</span>
<span class="c1"># can be found in the documentation for PilupEncoder class.</span>
<span class="n">encoding_layers</span> <span class="o">=</span> <span class="p">[</span><span class="n">PileupEncoder</span><span class="o">.</span><span class="n">Layer</span><span class="o">.</span><span class="n">READ</span><span class="p">,</span> <span class="n">PileupEncoder</span><span class="o">.</span><span class="n">Layer</span><span class="o">.</span><span class="n">BASE_QUALITY</span><span class="p">]</span>
<span class="n">pileup_encoder</span> <span class="o">=</span> <span class="n">PileupEncoder</span><span class="p">(</span>
    <span class="n">window_size</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">max_reads</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">layers</span><span class="o">=</span><span class="n">encoding_layers</span><span class="p">)</span>

<span class="c1"># Instantiate a zygosity encoder that generates output labels. Converts a variant entry</span>
<span class="c1"># into a class label for no variant, homozygous variant or heterozygous variant.</span>
<span class="n">zyg_encoder</span> <span class="o">=</span> <span class="n">ZygosityLabelEncoder</span><span class="p">()</span>

<span class="c1"># Create neural network that receives 2 channel inputs (encoding layers defined above)</span>
<span class="c1"># and outputs a logit over three classes (no variant, homozygous variant, heterozygous variant.</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">AlexNet</span><span class="p">(</span><span class="n">num_input_channels</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span>
    <span class="n">encoding_layers</span><span class="p">),</span> <span class="n">num_output_logits</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>

<span class="c1"># Get datasets to train on.</span>
<span class="c1"># NOTE: To train a neural network well, the model needs to see samples from all types of classes.</span>
<span class="c1"># The example here shows a file that has true variant (either homozygous or heterozygous),</span>
<span class="c1"># but in practice one also needs to pass a set of false positive samples so the model can learn to</span>
<span class="c1"># ignore them. False positive samples can be marked with `is_fp` so the reader can appripriately</span>
<span class="c1"># assign their variant types.</span>
<span class="n">bam</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">get_data_folder</span><span class="p">(),</span> <span class="s2">&quot;small_bam.bam&quot;</span><span class="p">)</span>
<span class="n">samples</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">get_data_folder</span><span class="p">(),</span> <span class="s2">&quot;candidates.vcf.gz&quot;</span><span class="p">)</span>
<span class="n">vcf_loader</span> <span class="o">=</span> <span class="n">VCFReader</span><span class="p">([</span><span class="n">VCFReader</span><span class="o">.</span><span class="n">VcfBamPath</span><span class="p">(</span><span class="n">vcf</span><span class="o">=</span><span class="n">samples</span><span class="p">,</span> <span class="n">bam</span><span class="o">=</span><span class="n">bam</span><span class="p">,</span> <span class="n">is_fp</span><span class="o">=</span><span class="kc">False</span><span class="p">)])</span>

<span class="c1"># Create a data loader with custom sample and label encoder.</span>
<span class="n">dataset_train</span> <span class="o">=</span> <span class="n">ReadPileupDataLoader</span><span class="p">(</span><span class="n">ReadPileupDataLoader</span><span class="o">.</span><span class="n">Type</span><span class="o">.</span><span class="n">TRAIN</span><span class="p">,</span> <span class="n">vcf_loader</span><span class="p">,</span>
                                     <span class="n">batch_size</span><span class="o">=</span><span class="mi">32</span><span class="p">,</span> <span class="n">shuffle</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                                     <span class="n">sample_encoder</span><span class="o">=</span><span class="n">pileup_encoder</span><span class="p">,</span> <span class="n">label_encoder</span><span class="o">=</span><span class="n">zyg_encoder</span><span class="p">)</span>

<span class="c1"># Use CrossEntropyLoss to train.</span>
<span class="n">vz_ce_loss</span> <span class="o">=</span> <span class="n">nemo</span><span class="o">.</span><span class="n">backends</span><span class="o">.</span><span class="n">pytorch</span><span class="o">.</span><span class="n">common</span><span class="o">.</span><span class="n">losses</span><span class="o">.</span><span class="n">CrossEntropyLossNM</span><span class="p">(</span><span class="n">logits_ndim</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

<span class="c1"># Create NeMo training DAG.</span>
<span class="n">vz_labels</span><span class="p">,</span> <span class="n">encoding</span> <span class="o">=</span> <span class="n">dataset_train</span><span class="p">()</span>
<span class="n">vz</span> <span class="o">=</span> <span class="n">model</span><span class="p">(</span><span class="n">encoding</span><span class="o">=</span><span class="n">encoding</span><span class="p">)</span>
<span class="n">vz_loss</span> <span class="o">=</span> <span class="n">vz_ce_loss</span><span class="p">(</span><span class="n">logits</span><span class="o">=</span><span class="n">vz</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">vz_labels</span><span class="p">)</span>

<span class="c1"># Logger callback</span>
<span class="n">logger_callback</span> <span class="o">=</span> <span class="n">nemo</span><span class="o">.</span><span class="n">core</span><span class="o">.</span><span class="n">SimpleLossLoggerCallback</span><span class="p">(</span>
    <span class="n">tensors</span><span class="o">=</span><span class="p">[</span><span class="n">vz_loss</span><span class="p">],</span>
    <span class="n">print_func</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Train Loss: </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">item</span><span class="p">())</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">))</span>
<span class="p">)</span>

<span class="c1"># Kick off training</span>
<span class="n">nf</span><span class="o">.</span><span class="n">train</span><span class="p">([</span><span class="n">vz_loss</span><span class="p">],</span>
         <span class="n">callbacks</span><span class="o">=</span><span class="p">[</span><span class="n">logger_callback</span><span class="p">,</span>
                    <span class="n">checkpoint_callback</span><span class="p">,</span> <span class="n">evaluator_callback</span><span class="p">],</span>
         <span class="n">optimization_params</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;num_epochs&quot;</span><span class="p">:</span> <span class="mi">4</span><span class="p">,</span> <span class="s2">&quot;lr&quot;</span><span class="p">:</span> <span class="mf">0.001</span><span class="p">},</span>
         <span class="n">optimizer</span><span class="o">=</span><span class="s2">&quot;adam&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="inference">
<h3>Inference<a class="headerlink" href="#inference" title="Permalink to this headline">¶</a></h3>
<p>The inference pipeline works in a very similar fashion, except the final NeMo DAG looks different.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import nemo and variantworks modules</span>
<span class="kn">import</span> <span class="nn">nemo</span>
<span class="kn">from</span> <span class="nn">variantworks.dataloader</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">variantworks.io.vcfio</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">variantworks.networks</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">variantworks.sample_encoders</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">variantworks.result_writer</span> <span class="kn">import</span> <span class="o">*</span>

<span class="c1"># Create neural factory. In this case, the checkpoint_dir has to be set for NeMo to pick</span>
<span class="c1"># up a pre-trained model.</span>
<span class="n">nf</span> <span class="o">=</span> <span class="n">nemo</span><span class="o">.</span><span class="n">core</span><span class="o">.</span><span class="n">NeuralModuleFactory</span><span class="p">(</span>
    <span class="n">placement</span><span class="o">=</span><span class="n">nemo</span><span class="o">.</span><span class="n">core</span><span class="o">.</span><span class="n">neural_factory</span><span class="o">.</span><span class="n">DeviceType</span><span class="o">.</span><span class="n">GPU</span><span class="p">,</span> <span class="n">checkpoint_dir</span><span class="o">=</span><span class="n">model_dir</span><span class="p">)</span>

<span class="c1"># Neural Network</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">AlexNet</span><span class="p">(</span><span class="n">num_input_channels</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span>
    <span class="n">encoding_layers</span><span class="p">),</span> <span class="n">num_output_logits</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>

<span class="c1"># Dataset generation is done in a similar manner. It&#39;s important to note that the encoder used</span>
<span class="c1"># for inference much match that for training.</span>
<span class="n">encoding_layers</span> <span class="o">=</span> <span class="p">[</span><span class="n">PileupEncoder</span><span class="o">.</span><span class="n">Layer</span><span class="o">.</span><span class="n">READ</span><span class="p">,</span> <span class="n">PileupEncoder</span><span class="o">.</span><span class="n">Layer</span><span class="o">.</span><span class="n">BASE_QUALITY</span><span class="p">]</span>
<span class="n">pileup_encoder</span> <span class="o">=</span> <span class="n">PileupEncoder</span><span class="p">(</span>
    <span class="n">window_size</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">max_reads</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">layers</span><span class="o">=</span><span class="n">encoding_layers</span><span class="p">)</span>

<span class="c1"># Similar to training, a dataloader needs to be setup for the relevant datasets. In the case of</span>
<span class="c1"># inference, it doesn&#39;t matter if the files are tagged as false positive or not. Each example will be</span>
<span class="c1"># evaluated by the network. For simplicity the example is using the same dataset from training.</span>
<span class="c1"># Note: No label encoder is required in inference.</span>
<span class="n">bam</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">test_data_dir</span><span class="p">,</span> <span class="s2">&quot;small_bam.bam&quot;</span><span class="p">)</span>
<span class="n">labels</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">test_data_dir</span><span class="p">,</span> <span class="s2">&quot;candidates.vcf.gz&quot;</span><span class="p">)</span>
<span class="n">vcf_bam_tuple</span> <span class="o">=</span> <span class="n">VCFReader</span><span class="o">.</span><span class="n">VcfBamPath</span><span class="p">(</span><span class="n">vcf</span><span class="o">=</span><span class="n">labels</span><span class="p">,</span> <span class="n">bam</span><span class="o">=</span><span class="n">bam</span><span class="p">,</span> <span class="n">is_fp</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">vcf_loader</span> <span class="o">=</span> <span class="n">VCFReader</span><span class="p">([</span><span class="n">vcf_bam_tuple</span><span class="p">])</span>
<span class="n">test_dataset</span> <span class="o">=</span> <span class="n">ReadPileupDataLoader</span><span class="p">(</span><span class="n">ReadPileupDataLoader</span><span class="o">.</span><span class="n">Type</span><span class="o">.</span><span class="n">TEST</span><span class="p">,</span> <span class="n">vcf_loader</span><span class="p">,</span> <span class="n">batch_size</span><span class="o">=</span><span class="mi">32</span><span class="p">,</span>
                                    <span class="n">shuffle</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">sample_encoder</span><span class="o">=</span><span class="n">pileup_encoder</span><span class="p">)</span>

<span class="c1"># Create inference DAG</span>
<span class="n">encoding</span> <span class="o">=</span> <span class="n">test_dataset</span><span class="p">()</span>
<span class="n">vz</span> <span class="o">=</span> <span class="n">model</span><span class="p">(</span><span class="n">encoding</span><span class="o">=</span><span class="n">encoding</span><span class="p">)</span>

<span class="c1"># Invoke the &quot;infer&quot; action.</span>
<span class="n">results</span> <span class="o">=</span> <span class="n">nf</span><span class="o">.</span><span class="n">infer</span><span class="p">([</span><span class="n">vz</span><span class="p">],</span> <span class="n">checkpoint_dir</span><span class="o">=</span><span class="n">model_dir</span><span class="p">,</span> <span class="n">verbose</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># Instantiate a decoder that converts the predicted output of the network to</span>
<span class="c1"># a zygosity enum.</span>
<span class="n">zyg_decoder</span> <span class="o">=</span> <span class="n">ZygosityLabelDecoder</span><span class="p">()</span>

<span class="c1"># Decode inference results to labels</span>
<span class="k">for</span> <span class="n">tensor_batches</span> <span class="ow">in</span> <span class="n">results</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">batch</span> <span class="ow">in</span> <span class="n">tensor_batches</span><span class="p">:</span>
        <span class="n">predicted_classes</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">batch</span><span class="p">,</span> <span class="n">dim</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">inferred_zygosity</span> <span class="o">+=</span> <span class="p">[</span><span class="n">zyg_decoder</span><span class="p">(</span><span class="n">pred</span><span class="p">)</span>
                             <span class="k">for</span> <span class="n">pred</span> <span class="ow">in</span> <span class="n">predicted_classes</span><span class="p">]</span>

<span class="c1"># Use the VCFResultWriter to output predicted zygosities to a VCF file.</span>
<span class="n">result_writer</span> <span class="o">=</span> <span class="n">VCFResultWriter</span><span class="p">(</span><span class="n">vcf_loader</span><span class="p">,</span> <span class="n">inferred_zygosity</span><span class="p">)</span>

<span class="n">result_writer</span><span class="o">.</span><span class="n">write_output</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Examples</a><ul>
<li><a class="reference internal" href="#snp-zygosity-predictor">SNP Zygosity Predictor</a><ul>
<li><a class="reference internal" href="#training">Training</a></li>
<li><a class="reference internal" href="#inference">Inference</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="features.html"
                        title="previous chapter">Core Features</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="apidocs/modules.html"
                        title="next chapter">API</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/examples.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="apidocs/modules.html" title="API"
             >next</a> |</li>
        <li class="right" >
          <a href="features.html" title="Core Features"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">VariantWorks SDK 0.1.0 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2020, NVIDIA Corportation.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 3.0.4.
    </div>
  </body>
</html>